### PR TITLE
fix:show index

### LINKED ISSUE #273: Bug: when execute "show index from xxx", return error.
**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

error:
```
mysql> show index from student;
ERROR 1146 (42S02): Table 'employees_0000.student' doesn't exist
mysql>
```

error log:
```
2022-07-03T17:11:30.411+0800    DEBUG   executor/redirect.go:161        ComQuery: show index from student
2022-07-03T17:11:30.411+0800    DEBUG   runtime/runtime.go:693  call upstream: db=employees_0000, sql="SHOW INDEXES FROM `student`", args=[]
2022-07-03T17:11:30.415+0800    ERROR   mysql/execute_handle.go:74      conn 0: error: Table 'employees_0000.student' doesn't exist (errno 1146) (sqlstate 42S02)
github.com/arana-db/arana/pkg/mysql.(*Listener).handleQuery.func1
        /Users/dongzonglei/source_code/Github/arana/pkg/mysql/execute_handle.go:74
github.com/arana-db/arana/pkg/mysql.(*Listener).handleQuery
        /Users/dongzonglei/source_code/Github/arana/pkg/mysql/execute_handle.go:102
github.com/arana-db/arana/pkg/mysql.(*Listener).ExecuteCommand
        /Users/dongzonglei/source_code/Github/arana/pkg/mysql/server.go:554
github.com/arana-db/arana/pkg/mysql.(*Listener).handle
        /Users/dongzonglei/source_code/Github/arana/pkg/mysql/server.go:203\
```

### PR DESCRIPTION
**Pull request by baerwang**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #273 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/runtime/ast/show.go [hunk 1]
```diff
@@ -220,6 +220,10 @@ func (s *ShowIndex) TableName() TableName {
 	return s.tableName
 }
 
+func (s *ShowIndex) SetTableName(tableName TableName) {
+	s.tableName = tableName
+}
+
```
Comment (dongzl) [inline pkg/runtime/ast/show.go:226]: Suggest make the `tableName` -> `TableName` directly, and remove this method.

**Review by dongzl — CHANGES_REQUESTED** [review]

File: pkg/runtime/plan/show_index.go [hunk 2]
```diff
@@ -49,14 +54,37 @@ func (s *ShowIndexPlan) ExecIn(ctx context.Context, conn proto.VConn) (proto.Res
...
-	var (
-		query = sb.String()
-		args  = s.toArgs(indexes)
+	query, err := conn.Query(ctx, db, sb.String(), s.toArgs(indexes)...)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	ds, _ := query.Dataset()
+	fields, _ := ds.Fields()
```
Comment (jjeffcaii) [inline pkg/runtime/plan/show_index.go:72]: Do not ignore error, should return err directly.

**Review by jjeffcaii — COMMENTED** [review]
fix ci

**Review by dongzl — APPROVED** [review]

File: pkg/runtime/optimize/optimizer.go [hunk 3]
```diff
@@ -857,8 +857,21 @@ func (o optimizer) optimizeShowTables(ctx context.Context, stmt *rast.ShowTables
...
 
-func (o optimizer) optimizeShowIndex(_ context.Context, stmt *rast.ShowIndex, args []interface{}) (proto.Plan, error) {
-	ret := &plan.ShowIndexPlan{Stmt: stmt}
+func (o optimizer) optimizeShowIndex(ctx context.Context, stmt *rast.ShowIndex, args []interface{}) (proto.Plan, error) {
+	var ru *rule.Rule
+	if ru = rcontext.Rule(ctx); ru == nil {
+		return nil, errors.WithStack(errNoRuleFound)
+	}
+
+	vt := ru.MustVTable(stmt.TableName.Suffix())
```
Comment (PangXing) [inline pkg/runtime/optimize/optimizer.go:866]: If table is non-sharding table,   it will panic.  I think it should be checked first.

Comment (baerwang) [reply pkg/runtime/optimize/optimizer.go:866]: this top have catch error

**Review by jjeffcaii — APPROVED** [review]
LGTM

**Review by AlexStocks — APPROVED** [review]

File: pkg/runtime/optimize/optimizer.go (hunk 3 above)
Comment (PangXing) [reply pkg/runtime/optimize/optimizer.go:866]: image
Without  your fix， non-sharding table can work , for example “show index from departments”. But now it cannot

**Review by PangXing — APPROVED** [review]

**Comment by dongzl** [comment]
image

**Pull request merged**